lib/ansiblelint/errors.py

```diff
@@ -18,20 +22,52 @@ def __repr__(self):
         return formatstr.format(self.rule.id, self.message,
                                 self.filename, self.linenumber, self.line)
 
+    def __lt__(self, other):
+        """Enable sorting of MatchError instances."""
+        return (
+                self.filename, self.linenumber, self.message, getattr(self.rule, "id", 0)
+            ) < (
+            other.filename, other.linenumber, other.message, getattr(other.rule, "id", 0))
 
+
+@functools.total_ordering
 class MatchError(ValueError):
     """Exception that would end-up as linter rule match."""
 
-    def __init__(self, message, rule=None):
+    def __init__(
+            self,
+            message: str = "",
+            rule: Any = None,
+            filename="",
+            linenumber=0,
+            line=None):
         """Initialize a MatchError instance."""
         super().__init__(message)
 
-        self.message = message
-        self.linenumber = 0
-        self.line = None
-        self.filename = None
+        if not (message or rule):
+            raise RuntimeError("Calling MatchError requires either a message or a rule.")
+
+        self.message = message or getattr(rule, "description", "")
+        self.linenumber = linenumber
+        self.line = line
+        self.filename = normpath(str(filename)) if filename else filename
```

Comment:
no need to stringify as it's done internally anyway
```suggestion
        self.filename = normpath(filename) if filename else filename
```